flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/expressions/CompositeTypeAccessExpressionITCase.java

```diff
@@ -0,0 +1,264 @@
+import org.junit.runners.Parameterized;
+import org.junit.runners.Suite;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static java.util.Collections.singletonMap;
+import static org.apache.flink.table.api.DataTypes.ARRAY;
+import static org.apache.flink.table.api.DataTypes.BIGINT;
+import static org.apache.flink.table.api.DataTypes.FIELD;
+import static org.apache.flink.table.api.DataTypes.MAP;
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.Expressions.$;
+import static org.apache.flink.table.api.Expressions.call;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for functions that access nested fields/elements of composite/collection types.
+ */
+@RunWith(Suite.class)
+@Suite.SuiteClasses(
+	{
+		CompositeTypeAccessExpressionITCase.TableFieldAccess.class,
+		CompositeTypeAccessExpressionITCase.CallFieldAccess.class
+	}
+)
+public class CompositeTypeAccessExpressionITCase {
+
+	/**
+	 * Regular tests. See also {@link CallFieldAccess} for tests that access a nested field of an expression or
+	 * for {@link BuiltInFunctionDefinitions#FLATTEN} which produces multiple columns from a single one.
+	 */
+	public static class TableFieldAccess extends BuiltInFunctionTestBase {
+		@Parameterized.Parameters(name = "{index}: {0}")
+		public static List<TestSpec> testData() {
+			return Arrays.asList(
+
+				// Actually in case of SQL it does not use the GET method, but
+				// a custom logic for accessing nested fields of a Table.
+				TestSpec.forFunction(BuiltInFunctionDefinitions.GET)
+					.onFieldsWithData(null, Row.of(1))
+					.andDataTypes(
+						ROW(FIELD("nested", BIGINT().notNull())).nullable(),
+						ROW(FIELD("nested", BIGINT().notNull())).notNull()
+					)
+					.testTableApiResult($("f0").get("nested"), null, BIGINT().nullable())
+					.testTableApiResult($("f1").get("nested"), 1L, BIGINT().notNull())
+					.testSqlResult("f0.nested", null, BIGINT().nullable())
+					.testSqlResult("f1.nested", 1L, BIGINT().notNull()),
+
+				// In Calcite it maps to FlinkSqlOperatorTable.ITEM
+				TestSpec.forFunction(BuiltInFunctionDefinitions.AT)
+					.onFieldsWithData(null, new int[] {1}, null, singletonMap("nested", 1), null, Row.of(1))
+					.andDataTypes(
+						ARRAY(BIGINT().notNull()).nullable(),
+						ARRAY(BIGINT().notNull()).notNull(),
+						MAP(STRING(), BIGINT().notNull()).nullable(),
+						MAP(STRING(), BIGINT().notNull()).notNull(),
+						ROW(FIELD("nested", BIGINT().notNull())).nullable(),
+						ROW(FIELD("nested", BIGINT().notNull())).notNull()
+					)
+					// accessing elements of MAP or ARRAY is a runtime operations,
+					// we do not know about the size or contents during the inference
+					// therefore the results are always nullable
+					.testSqlResult("f0[1]", null, BIGINT().nullable())
```

Comment:
should we port `AT` as well or in a separate PR?

Reply:
I have it in a different branch. I found this PR already quite big.